core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala

```diff
@@ -0,0 +1,223 @@
+        "stderr for pgrep",
+        process.getErrorStream,
+        line => errorStringBuilder.append(line))
+      val exitCode = process.waitFor()
+      stdoutThread.join()
+      stdErrThread.join()
+      val errorString = errorStringBuilder.toString()
+      // pgrep will have exit code of 1 if there are more than one child process
+      // and it will have a exit code of 2 if there is no child process
+      if (exitCode != 0 && exitCode > 2) {
+        val cmd = builder.command().toArray.mkString(" ")
+        logWarning(s"Process $cmd exited with code $exitCode and stderr: $errorString")
+        throw new SparkException(s"Process $cmd exited with code $exitCode")
+      }
+      childPidsInInt
+    } catch {
+      case e: Exception =>
+        logWarning("Exception when trying to compute process tree." +
+          " As a result reporting of ProcessTree metrics is stopped.", e)
+        isAvailable = false
+        mutable.ArrayBuffer.empty[Int]
+    }
+  }
+
+  def addProcfsMetricsFromOneProcess(
+      allMetrics: ProcfsMetrics,
+      pid: Int): ProcfsMetrics = {
+
+    // The computation of RSS and Vmem are based on proc(5):
+    // http://man7.org/linux/man-pages/man5/proc.5.html
+    try {
+      val pidDir = new File(procfsDir, pid.toString)
+      Utils.tryWithResource(new InputStreamReader(
+        new FileInputStream(
+          new File(pidDir, procfsStatFile)), Charset.forName("UTF-8"))) { fReader =>
+          val in = new BufferedReader(fReader)
+          val procInfo = in.readLine
+          val procInfoSplit = procInfo.split(" ")
+          val vmem = procInfoSplit(22).toLong
+          val rssMem = procInfoSplit(23).toLong * pageSize
+          if (procInfoSplit(1).toLowerCase(Locale.US).contains("java")) {
+            allMetrics.copy(
+              jvmVmemTotal = allMetrics.jvmVmemTotal + vmem,
+              jvmRSSTotal = allMetrics.jvmRSSTotal + (rssMem)
+            )
+          }
+          else if (procInfoSplit(1).toLowerCase(Locale.US).contains("python")) {
+            allMetrics.copy(
+              pythonVmemTotal = allMetrics.pythonVmemTotal + vmem,
+              pythonRSSTotal = allMetrics.pythonRSSTotal + (rssMem)
+            )
+          }
+          else {
+            allMetrics.copy(
+              otherVmemTotal = allMetrics.otherVmemTotal + vmem,
+              otherRSSTotal = allMetrics.otherRSSTotal + (rssMem)
+            )
+          }
+      }
+    } catch {
+      case f: IOException =>
+        logWarning("There was a problem with reading" +
+          " the stat file of the process. ", f)
+        ProcfsMetrics(0, 0, 0, 0, 0, 0)
+    }
+  }
+
+  private[spark] def computeAllMetrics(): ProcfsMetrics = {
+    if (!isAvailable) {
+      return ProcfsMetrics(0, 0, 0, 0, 0, 0)
+    }
+    val pids = computeProcessTree
+    var allMetrics = ProcfsMetrics(0, 0, 0, 0, 0, 0)
+    for (p <- pids) {
+      allMetrics = addProcfsMetricsFromOneProcess(allMetrics, p)
+    }
+    allMetrics
```

Comment:
one weird thing about the error handling here -- if you get any error from `computeProcessTree` through all calls to `addProcsFsMetricsFromOneProcess`, then you'll set `isAvailable = false`, but you'll still return `allMetrics`.  Depending on when the error occurs, you might have some partially accumulated metrics in there.  I think if there is any error, you probably want to return empty metrics.

Also you'll leave `isAvailable=false` forever after that -- is that OK?  I guess I don't really have strong feelings about this one, maybe its OK to give up forever for this executor.

Reply:
It is possible to be more accurate here when an error happens. Although I think I need to revert to some older version of method signature, since right now there is a cohesion around allMetrics. I will do something tomorrow.
I don't want to introduce overhead of keep checking, but I can think about it and implement some sort of a check after we set isAvailable = false because of an error.

Reply:
My comment about overhead was referring to your question about whether is it ok to keep `isAvailable=false` after error. I guess we can go with the option of not changing isAvailable= false after an error in computing processTree and just set a flag to check afterward and return zero in case of error like the way you suggested in your above comment. That way we will not stop reporting for this executor and we only return zero in case of error. I'm fine with this as well.